### DISCUSSION TITLE
Multiple prisma clients and databases

### DISCUSSION THREAD

**Discussion by emilbryggare** [vercel/turborepo#3493] [Ideas]
Using Prisma with Turborepo works great when following https://turbo.build/repo/docs/handbook/tools/prisma if there is only one Prisma client in the monorepo.

However, when running Turborepo with multiple Prisma clients and databases they will overwrite each other. It is OK if you only use one database per app, then you would just have to be careful when you run `prisma generate` so the client is generated for the app you are working on.

But if you have multiple clients/databases in one app things get more difficult.  There are loads of hacks out there, some mentioned here [prisma/prisma#10433 (comment)](https://github.com/prisma/prisma/issues/10433#issuecomment-1252089060) but no real solutions as far as I can tell. 

I've added a simple reproduction of the use case based on with-prisma example here: https://github.com/emilbryggare/turbo/pull/1

Just wanted to raise awareness of this here and maybe it is worth adding a caveat to the docs about this so users are aware. What do you think?

**Comment by Neosoulink** [comment]
Hi @emilbryggare, did you find out how to solve that?

I have my turbo repo project manages with `pnpm` and inside of it i do have one package containing a main database, but also inside two apps in the project they have their own __Prisma__ schema.

Yes I'm getting a very very weird behavior when trying to generate codes, because of __symlinks__ pnpm strategy, my generated codes are overwriting each others and I'm not able to use my __Prisma__ logics correctly

**Reply by emilbryggare** [reply]
I did not manage to solve it back then, but have not had the need to look into it more since then. Sorry.

**Reply by Neosoulink** [reply]
Hey @emilbryggare, sorry to come back to you like 3 weeks later 😅

But I solved the problem. I specified the `output` directory inside all the `prisma.schema` files inside my turbo-repo project like follow:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = []
  output          = "../node_modules/.prisma/client"
}
```
> 💡 Note that the node_modules in the above output directory represent the app/package node_modules

And I changed the default import from `@prisma/` to `.prisma/`

E.g: `import { PrismaClient } from ".prisma/client";`

**Reply by srizvi** [reply]
Thanks for the solution @Neosoulink, worked for me.

**Reply by adamspotlite** [reply]
@Neosoulink did you have any problems with seed scripts for this? getting:

```
TypeError [ERR_INVALID_MODULE_SPECIFIER]: Invalid module ".prisma/client" is not a valid package name imported from /home/some-project/packages/db/src/index.ts
```

even tho typescript seems fine...

**Reply by Neosoulink** [reply]
Hum... Unfortunately, I did not have that problem back then, and all my seeders were working just fine.

Is it possible to share more details (or repo link) for your issue?

It looks like this one is not a problem related to turbo-repo or prisma

**Reply by adamspotlite** [reply]
i pulled out the seed script into a different package, so that i could `import { prisma } from "@acme/first-prisma-package"`, hacking my way around the issue :sweat_smile: 

Will try to get back around to a repo link in the next couple days :+1:

**Comment by anthonyshew** [comment]
Looks like @Neosoulink's answer is working for folks, judging by emoji reactions!

> Hey @emilbryggare, sorry to come back to you like 3 weeks later 😅
> 
> But I solved the problem. I specified the `output` directory inside all the `prisma.schema` files inside my [turbo-repo](https://linear.app/vercel/issue/TURBO-repo) project like follow:
> 
> ```prisma
> generator client {
>   provider        = "prisma-client-js"
>   previewFeatures = []
>   output          = "../node_modules/.prisma/client"
> }
> ```
> 
> > 💡 Note that the node_modules in the above output directory represent the app/package node_modules
> 
> And I changed the default import from `@prisma/` to `.prisma/`
> 
> E.g: `import { PrismaClient } from ".prisma/client";`

In the interest of cleaning up our GitHub Discussions, I'll mark this as closed! (I would mark it as answered but the GitHub UI isn't letting me do so.) 😄

**Reply by jmunozz** [reply]
This is not working for me as I come by this error when running the built version: 

```
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_MODULE_SPECIFIER]: Invalid module ".prisma/client" is not a valid package name imported from /Users/ledjodje/projects/librairie-theatrale/noticit/apps/diliwhat/dist/db/client.js
    at new NodeError (node:internal/errors:399:5)
    at parsePackageName (node:internal/modules/esm/resolve:722:11)
    at packageResolve (node:internal/modules/esm/resolve:745:5)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/hooks:654:28)
    at Hooks.resolve (node:internal/modules/esm/hooks:309:30)
    at ESMLoader.resolve (node:internal/modules/esm/loader:312:26)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:172:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40) {
  code: 'ERR_INVALID_MODULE_SPECIFIER'
}

Node.js v19.8.1
```
I'm using ESM.

**Reply by Neosoulink** [reply]
Hey @jmunozz , can you share more details ?

Is there a public repo to check?

**Reply by jmunozz** [reply]
No public repo. 
Basically I used your solution in a turborepo project where I need each app to have its own database with its own schema. 
When running `npx prisma generate` it created a `node_modules/.prisma` folder with this package.json inside: 
```
{
  "name": ".prisma/client",
  "main": "index.js",
  "types": "index.d.ts",
  "browser": "index-browser.js",
  "sideEffects": false
}
```
I then built my app. No problem yet. 
But when I ran `dist/index.js`, the error happened.

I am using a specific tsconfig with  and ESM module in my package.json. 
Following is my tsconfig.json: 
```
{
    "$schema": "https://json.schemastore.org/tsconfig",
    "display": "Default",
    "compilerOptions": {
      "esModuleInterop": true, 
      "module": "NodeNext", 
      "target": "esnext", 
      "sourceMap": true, 
      "strict": true, 
      "composite": false,
      "declaration": true,
      "declarationMap": true,
      "forceConsistentCasingInFileNames": true,
      "inlineSources": false,
      "isolatedModules": true,
      "moduleResolution": "NodeNext",
      "noUnusedLocals": false,
      "noUnusedParameters": false,
      "preserveWatchOutput": true,
      "skipLibCheck": true,
      "strictNullChecks": true
    },
    "exclude": ["node_modules", "__tests__"]
}
```
I think the probleme came from the fact that I tried to import it as a package. 
I managed to fix it by replacing output and import it as a I would do from any other file in my project.

```
generator client {
    provider = "prisma-client-js"
    output   = "./client"
}
```

Then in `db/client.ts`

```
import { PrismaClient } from "../../prisma/client/index.js";
export * from "../../prisma/client/index.js";

declare global {
  var prisma: PrismaClient | undefined;
}

export const prisma = global.prisma || new PrismaClient();

if (process.env.NODE_ENV !== "production") global.prisma = prisma;
```

**Reply by Neosoulink** [reply]
@jmunozz, yes, this is another approach to solving the problem... Even though you lose the packaging logic by doing that

The idea is just to avoid overwriting the Prisma-generated code.

--- 
> Maybe you did that already, but don't forget to add the folder `./client/` in you `.gitignore`

**Reply by adamspotlite** [reply]
@jmunozz i had the same thing come about, although originally I didn't. Did you end up staying with this solution?

**Reply by ManupaDev** [reply]
This works fine locally and even build on Vercel but when i do visit a to a page with a db fetch it fails with the follwoing error

`Unhandled Rejection: PrismaClientInitializationError: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x". We detected that you are using Next.js, learn how to fix this: https://pris.ly/d/engine-not-found-nextjs. This is likely caused by a bundler that has not copied "libquery_engine-rhel-openssl-3.0.x.so.node" next to the resulting bundle. Ensure that "libquery_engine-rhel-openssl-3.0.x.so.node" has been copied next to the bundle or in "node_modules/.prisma/client". We would appreciate if you could take the time to share some information with us. Please help us by answering a few questions: https://pris.ly/engine-not-found-bundler-investigation The following locations have been searched: /var/task/apps/web/node_modules/.prisma/client /var/task/apps/web/.next/server /vercel/path0/apps/web/node_modules/.prisma/client /var/task/apps/web/.prisma/client /tmp/prisma-engines at wa (/var/task/apps/web/.next/server/chunks/5042.js:64:756) at async Object.loadLibrary (/var/task/apps/web/.next/server/chunks/5042.js:111:10016) at async wt.loadEngine (/var/task/apps/web/.next/server/chunks/5042.js:112:448) at async wt.instantiateLibrary (/var/task/apps/web/.next/server/chunks/5042.js:111:12736) { clientVersion: '5.15.1', errorCode: undefined } Node.js process exited with exit status: 128. The logs above can help with debugging the issue. `

**Reply by jharrell** [reply]
@ManupaDev hello! I'm from the Prisma team and we're looking into this. 

These problems _should_ be resolved if you use our [monorepo plugin](https://www.npmjs.com/package/@prisma/nextjs-monorepo-workaround-plugin) ([docs available here](https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-monorepo))

If that's not working, we'd love a reproduction so we can figure out what's going on. Would that be possible?

**Comment by phuctyouup** [comment]
You guys are banging your head against a wall breaking every rule for a sustainable solid working turbo monorepo + Prisma. What your are building wont last. wont pass any security or audit tests. 
the closest you can get would be;

###    Add multiple schema names to the datasource:    
                                                                                                                                                
```
   datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["app1", "app2", "shared"]
```

### Assign models to specific schemas:

```
model User {
  id Int @id
  // ...
  @@schema("app1")
}

model Product {
  id Int @id
  // ...
  @@schema("app2")
}
```
This lets one Prisma Client query across schemas in the same database, with relations possible across them.
It's not for completely separate databases.

Prisma does not and can not properly work with completely seperate DB's. You cant do it. you can write workarounds till your little fingers fall off. and it will seem to work, but data integrity will NEVER, NEVER be stable. NEVER. you will have rules overwritting rules until it doesnt know what to do and you will start getting just a bunch of incoherent jarbled duplicated data popping up in places that you wont know how it got there and unable to trace it back to its origin. you wont be able to fix it when it breaks. and it will eventually break. and it will be impossible to trace to its source because its the accumulated rules bypassing rules to get the wanted results. which works at first but will eventually fail.

 ****@jharrell with the prisma team. Hows that coming? Let me know when you get done working on that. God speed to you and your team.**** 

Yours Truly
Phuctyououp

**Reply by jharrell** [reply]
@phuctyouup  thanks for your comment and your passion. I’m no longer with Prisma, I’d recommend asking on their Twitter or GitHub

**Reply by phuctyouup** [reply]
what was your position on the team and how long were you with the prisma team? asking because i will be looking to hire some people from prisma teams on a start up im working on will be in first round of funding within the year. I need experts, well experts is a very hard label to put on something. most people dont consider themselves experts, maybe a better definition would be a well weathered prisma employee who has a very strong background and firm understanding of current turbo monorepos hybrid models with a solid understanding of data retrieval methods and concepts. ill probably leave out hybrid models there probably wont be many or any of those. but it is still going to be a hard onboarding for anyone to get used to these new rules and schemas and certain coding 60-70% familiarity with 10-30% rethinking everything you know and take as correct aand conical. some wont ever grasp others might adapt sooner than others depends on how fast you can make the new connections. some of it will just seem like magic and might never really understand how it works and just learns the code and very strict rules it they follow the rules and dont do the old way they will be fine i guess you really dont know how it works, but the engineer minds they will not move forward untilit makes since to them and if they cant grasp and make the connections they will never adapt and they will go back but the ones who grasp and understand and really know that the magic really is magic. its not magic but it sures seems lie it at first your mind goes bonkers and you think you might just be as stupid as some people have told you you were but once that clicks you will have a what the fuck moment and be fine. so seriously let me know what you did over ther.

Best Regards 

Phuct (for now)
sry about the type Oh's but i never use any correction and refuse to go back and fix my appendages mess ups. they can do it themselves.
